core/cpuid.c

```diff
@@ -114,3 +131,234 @@ bool cpuid_host_has_feature_uncached(uint32_t feature_key)
+{
+    *cpuid_1_features_ecx         = cpuid->feature_1_ecx;
+    *cpuid_1_features_edx         = cpuid->feature_1_edx;
+    *cpuid_8000_0001_features_ecx = cpuid->feature_8000_0001_ecx;
+    *cpuid_8000_0001_features_edx = cpuid->feature_8000_0001_edx;
+}
+
+void cpuid_set_guest_features(cpuid_t *cpuid, hax_cpuid *cpuid_info)
+{
+    static void (*cpuid_set_guest_feature[])(cpuid_t *, hax_cpuid *) = {
+        cpuid_set_0000_0001,
+        cpuid_set_8000_0001
+    };
+    static size_t count = sizeof(cpuid_set_guest_feature) /
+                          sizeof(cpuid_set_guest_feature[0]);
+
+    hax_log(HAX_LOGI, "%s: before:\n", __func__);
+    hax_log(HAX_LOGI, "feature_1_ecx: %08lx, feature_1_edx: %08lx\n",
+            cpuid->feature_1_ecx, cpuid->feature_1_edx);
+    hax_log(HAX_LOGI, "feature_8000_0001_ecx: %08lx, feature_8000_0001_edx: %08lx"
+            "\n", cpuid->feature_8000_0001_ecx, cpuid->feature_8000_0001_edx);
+
+    for (int i = 0; i < count; ++i) {
+        cpuid_set_guest_feature[i](cpuid, cpuid_info);
+    }
+
+    hax_log(HAX_LOGI, "%s: after:\n", __func__);
+    hax_log(HAX_LOGI, "feature_1_ecx: %08lx, feature_1_edx: %08lx\n",
+            cpuid->feature_1_ecx, cpuid->feature_1_edx);
+    hax_log(HAX_LOGI, "feature_8000_0001_ecx: %08lx, feature_8000_0001_edx: %08lx"
+            "\n", cpuid->feature_8000_0001_ecx, cpuid->feature_8000_0001_edx);
+}
+
+static hax_cpuid_entry * find_cpuid_entry(hax_cpuid *cpuid_info,
+                                          uint32_t function, uint32_t index)
+{
+    int i;
+    hax_cpuid_entry *entry, *found = NULL;
+
+    for (i = 0; i < cpuid_info->total; ++i) {
+        entry = &cpuid_info->entries[i];
+        if (entry->function == function && entry->index == index) {
+            found = entry;
+            break;
+        }
+    }
+
+    return found;
+}
+
+static void cpuid_set_0000_0001(cpuid_t *cpuid, hax_cpuid *cpuid_info)
+{
+    const uint32_t kFunction = 0x01;
+    hax_cpuid_entry *entry;
+
+    entry = find_cpuid_entry(cpuid_info, kFunction, 0);
+    if (entry == NULL)
+        return;
+
+    hax_log(HAX_LOGI, "%s: function: %08lx, index: %lu, flags: %08lx\n",
+            __func__, entry->function, entry->index, entry->flags);
+    hax_log(HAX_LOGI, "%s: eax: %08lx, ebx: %08lx, ecx: %08lx, edx: %08lx\n",
+            __func__, entry->eax, entry->ebx, entry->ecx, entry->edx);
+
+    cpuid->feature_1_ecx = entry->ecx;
+    cpuid->feature_1_edx = entry->edx;
+
+    // Filter the unsupported features
+    cpuid->feature_1_ecx &= cache.host_supported.feature_1_ecx &
+                            cache.hax_supported.feature_1_ecx;
+    cpuid->feature_1_edx &= cache.host_supported.feature_1_edx &
+                            cache.hax_supported.feature_1_edx;
+
+    // Set fixed supported features
+    cpuid_set_fixed_features(cpuid);
+
+    if (entry->ecx != cpuid->feature_1_ecx ||
+        entry->edx != cpuid->feature_1_edx) {
```

Comment:
There are 3 types of action with the actual supported features and the ones provided in ioctl:
1) Use intersection and ignore mismatch. This is currently used.
2) Use intersection and return error in case of mismatch. The user wanted the provided features and we can't support these.
3) Use provided features, no intersection with supported. It is users responsibility when the provided features are not supported by the cpu. This is the case of KVM, but it supports a lot of features, so the problems will be in case when the CPU does not support them.
Would be good add this as flags to the interface, at least in future.

Reply:
Thanks for your suggestion. We will consider these cases in future.